### DISCUSSION TITLE
How can authenticated user update its own user fields?

### DISCUSSION THREAD

**Discussion by YarivGilad** [strapi/strapi#7018] [Legacy Topics]
After authentication, how can the client app update the user model fields?
where can I find the user routes?

**Comment by CtFelix** [comment]
If i correctly understand your question, you should give users permission to update profile
> Roles & Permissions > USERS-PERMISSIONS > User > Update

Then use endpoint 

> PUT /users/:id

JSON
```
{
    "Username": "SomeNewName",
    "WhatEverYouNeed": "SomeData"
}
```

**Reply by arutr** [reply]
This is for updating any user record, which is probably not what @YarivGilad is looking for

**Reply by sandiprb** [reply]
Yes, @arutr  is correct. 

@YarivGilad you can create a custom controller 

with route config of `user/:id/update`

in your controller, you can do as below

```
updateUserProfile: async (ctx) => {
    const userIdToUpdate = ctx.params.id;
    const currentUserId = ctx.state.user.id;

    if (userIdToUpdate !== currentUserId) {
      ctx.status = 403;
      ctx.body = "Action not allowed";
      return; 
    }

// continue if the userIdToUpdate is of current user 
    try {
      const userData = await strapi
        .query("user", "users-permissions")
        .findOne({ id: userIdToUpdate });

      // your update logic goes here

      return  // your updated user data or custom response 
    } catch (error) {
      // send an error response
    }
  },
```

**Reply by derrickmehaffy** [reply]
The option from @sandiprb is a good option yes and is what I would recommend. I would say to make sure the user can't change fields they shouldn't have access to (such as changing their own role, or blocking/unblocking themselves)

**Comment by nitewulf** [comment]
I've never gotten this far, but I believe the answer would be to enable that permission but create a custom controller which would validate that the user making the request is only updating their user profile or is not an admin which would be allowed to update anyone, if you wanted.
